File: apps/native/services/appMetadata.ts
import Constants from "expo-constants";
import { getApiUrl } from "../firebase.config";

export interface VersionConfig {
  latestVersion: string;
  minVersion: string;
  updateType: "mandatory" | "optional";
  forceUpdateMessage?: string;
  optionalUpdateMessage?: string;
}

export interface MaintenanceMode {
  enabled: boolean;
  message: string;
  allowedVersions?: string[];
}

export interface StoreUrls {
  ios: string;
  android: string;
}

export interface SupportUrls {
  supportEmail: string;
  supportUrl: string;
  termsUrl: string;
  privacyUrl: string;
}

export interface AppMetadata {
  versionConfig: VersionConfig | null;
  featureFlags: Record<string, boolean>;
  maintenanceMode: MaintenanceMode | null;
  storeUrls: StoreUrls | null;
  supportUrls: SupportUrls | null;
}

export type UpdateStatus = "up_to_date" | "update_available" | "update_required";

export function getAppVersion(): string {
  return Constants.expoConfig?.version || "1.0.0";
}

/**
 * Compare two semver strings. Returns:
 *  -1 if a < b, 0 if a == b, 1 if a > b
 */
function compareSemver(a: string, b: string): number {
  const pa = a.split(".").map(Number);
  const pb = b.split(".").map(Number);
  for (let i = 0; i < 3; i++) {
    const na = pa[i] || 0;
    const nb = pb[i] || 0;
    if (na < nb) return -1;
    if (na > nb) return 1;
  }
  return 0;
}

export function getUpdateStatus(
  currentVersion: string,
  versionConfig: VersionConfig | null,
): UpdateStatus {
  if (!versionConfig) return "up_to_date";

  const { latestVersion, minVersion, updateType } = versionConfig;

  // Below minimum version — always required
  if (compareSemver(currentVersion, minVersion) < 0) {
    return "update_required";
  }

  // Below latest version
  if (compareSemver(currentVersion, latestVersion) < 0) {
    return updateType === "mandatory" ? "update_required" : "update_available";
  }

  return "up_to_date";
}

export async function fetchAppMetadata(): Promise<AppMetadata> {
  const apiUrl = getApiUrl();
  const response = await fetch(`${apiUrl}/config/app-metadata`);

  if (!response.ok) {
    throw new Error(`Failed to fetch app metadata: ${response.status}`);
  }

  const data = await response.json();

  return {
    versionConfig: (data.version_config as VersionConfig) || null,
    featureFlags: (data.feature_flags as Record<string, boolean>) || {},
    maintenanceMode: (data.maintenance_mode as MaintenanceMode) || null,
    storeUrls: (data.store_urls as StoreUrls) || null,
    supportUrls: (data.support_urls as SupportUrls) || null,
  };
}

File: apps-native/flutter-app/lib/core/config/app_metadata.dart
/// Dart models mirroring the `app_config` registry shapes defined in
/// `packages/api/src/types/config.ts`. These are consumed from
/// `GET /api/config/app-metadata`.
library app_metadata;

enum UpdateType {
  mandatory,
  optional;

  static UpdateType fromWire(String value) {
    return UpdateType.values.firstWhere(
      (v) => v.name == value,
      orElse: () => UpdateType.optional,
    );
  }
}

class VersionConfig {
  const VersionConfig({
    required this.latestVersion,
    required this.minVersion,
    required this.updateType,
    this.forceUpdateMessage,
    this.optionalUpdateMessage,
  });

  final String latestVersion;
  final String minVersion;
  final UpdateType updateType;
  final String? forceUpdateMessage;
  final String? optionalUpdateMessage;

  factory VersionConfig.fromJson(Map<String, dynamic> json) {
    return VersionConfig(
      latestVersion: json['latestVersion'] as String,
      minVersion: json['minVersion'] as String,
      updateType: UpdateType.fromWire(json['updateType'] as String),
      forceUpdateMessage: json['forceUpdateMessage'] as String?,
      optionalUpdateMessage: json['optionalUpdateMessage'] as String?,
    );
  }
}

/// `Record<string, boolean>` on the TS side.
typedef FeatureFlags = Map<String, bool>;

FeatureFlags featureFlagsFromJson(Map<String, dynamic> json) {
  return json.map((key, value) => MapEntry(key, value as bool));
}

class MaintenanceMode {
  const MaintenanceMode({
    required this.enabled,
    required this.message,
    this.allowedVersions,
  });

  final bool enabled;
  final String message;
  final List<String>? allowedVersions;

  factory MaintenanceMode.fromJson(Map<String, dynamic> json) {
    return MaintenanceMode(
      enabled: json['enabled'] as bool,
      message: json['message'] as String,
      allowedVersions: (json['allowedVersions'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList(),
    );
  }
}

class StoreUrls {
  const StoreUrls({required this.ios, required this.android});

  final String ios;
  final String android;

  factory StoreUrls.fromJson(Map<String, dynamic> json) {
    return StoreUrls(
      ios: json['ios'] as String,
      android: json['android'] as String,
    );
  }
}

class SupportUrls {
  const SupportUrls({
    required this.supportEmail,
    required this.supportUrl,
    required this.termsUrl,
    required this.privacyUrl,
  });

  final String supportEmail;
  final String supportUrl;
  final String termsUrl;
  final String privacyUrl;

  factory SupportUrls.fromJson(Map<String, dynamic> json) {
    return SupportUrls(
      supportEmail: json['supportEmail'] as String,
      supportUrl: json['supportUrl'] as String,
      termsUrl: json['termsUrl'] as String,
      privacyUrl: json['privacyUrl'] as String,
    );
  }
}

/// Response shape of `GET /api/config/app-metadata`. All keys are
/// optional — an unset key means the server has no override for it.
class AppMetadata {
  const AppMetadata({
    this.versionConfig,
    this.featureFlags,
    this.maintenanceMode,
    this.storeUrls,
    this.supportUrls,
  });

  final VersionConfig? versionConfig;
  final FeatureFlags? featureFlags;
  final MaintenanceMode? maintenanceMode;
  final StoreUrls? storeUrls;
  final SupportUrls? supportUrls;

  factory AppMetadata.fromJson(Map<String, dynamic> json) {
    return AppMetadata(
      versionConfig: json['version_config'] != null
          ? VersionConfig.fromJson(json['version_config'] as Map<String, dynamic>)
          : null,
      featureFlags: json['feature_flags'] != null
          ? featureFlagsFromJson(json['feature_flags'] as Map<String, dynamic>)
          : null,
      maintenanceMode: json['maintenance_mode'] != null
          ? MaintenanceMode.fromJson(json['maintenance_mode'] as Map<String, dynamic>)
          : null,
      storeUrls: json['store_urls'] != null
          ? StoreUrls.fromJson(json['store_urls'] as Map<String, dynamic>)
          : null,
      supportUrls: json['support_urls'] != null
          ? SupportUrls.fromJson(json['support_urls'] as Map<String, dynamic>)
          : null,
    );
  }
}
